vmdb/spec/models/authenticator/amazon_spec.rb

```diff
@@ -0,0 +1,461 @@
+require "spec_helper"
+require "aws-sdk"
+
+describe Authenticator::Amazon do
+  subject { Authenticator::Amazon.new(config) }
+  let!(:alice) { FactoryGirl.create(:user, :userid => 'alice') }
+  let(:config) do
+    {
+      :amazon_role   => false,
+      :amazon_key    => 'awskey',
+      :amazon_secret => 'awssecret',
+    }
+  end
+
+  class FakeAmazon
+    def initialize(user_data, *connect_args)
+      @user_data = user_data
+      _connect *connect_args
+    end
+
+    def _connect(username, password, service = nil)
+      if @user_data[username] && @user_data[username][:password] == password
+        @user = username
+      else
+        raise AWS::EC2::Errors::AuthFailure
+      end
+    end
+
+    def regions
+      []
+    end
+
+    def users
+      @user_data.reject { |k, _v| k =~ /aws/ }.map { |k, v| FakeUser.new(k, v) }
+    end
+
+    def client
+      iam_username =
+        case @user_data[@user][:iam]
+        when false
+          nil
+        when :denied
+          raise AWS::IAM::Errors::AccessDenied
+        else
+          @user
+        end
+
+      Object.new.tap do |obj|
+        class << obj
+          attr_accessor :username
+          def get_user
+            {:user => {:user_name => username}}
+          end
+        end
+        obj.username = iam_username
+      end
+    end
+
+    class FakeUser < Struct.new(:username, :data)
+      def name
+        data[:name]
+      end
+
+      def access_keys
+        data[:access_keys].map { |key| FakeKey.new(key) }
+      end
+
+      def groups
+        data[:groups].map { |group| FakeGroup.new(group) }
+      end
+    end
+
+    class FakeKey < Struct.new(:id)
+    end
+
+    class FakeGroup < Struct.new(:name)
+    end
+  end
+
+  before(:each) do
+    # If anything goes looking for the currently configured
+    # Authenticator during any of these tests, we'd really rather they
+    # found the one we're working on.
+    #
+    # This specifically comes up when we auto-create a new user from an
+    # external auth system: they get saved without a password, so User's
+    # dummy_password_for_external_auth hook runs, and it needs to ask
+    # Authenticator#password? whether it's allowed to do anything.
+
+    allow(User).to receive(:authenticator).and_return(subject)
+  end
+
+  before(:each) do
```

Comment:
I slightly prefer the literate flow of "before each do" -- but am happy to bend to precedent/preference.

Two blocks: because the below 6 lines are all doing one thing, which is unrelated to the one line above... and said line carries a (rather long-winded) comment. This way, they're clearly two very separate things, that both happen to occur at the same point in the test-running lifecycle.